### PR TITLE
MISSION, TOOLS, GAZEBO

### PR DESCRIPTION
**Pull request by mattlangford**
This was more an adaptation rather than a port. 
- The mission system is still mainly code from last year, with some additions that were based on Navigator, mainly the yaml config for vision services. There are some rules about what your perception nodes need to include that are worth looking at in the comments in the main `sub_singleton` file.
- Rviz config! It's pretty new and I only used it in sim so there could be more things we need to show on there.
- fprint! Everyone's favorite printing method. I'm going to make some more commits that transition a lot of the old nodes to use fprint.
- `sub8_ros_tools` and `sub8_misc_tools` used to require you to import both to use tools in both. Now all tools from both can be imported by importing `sub8_tools`.
- Transitioned the height over bottom kill to be an alarm kill not a kill handling kill

Also, here are the aliases I have:
```bash
alias sub="cd ~/mil_ws/src/Sub8"
alias subm="rosrun sub8_missions run_mission"
alias subc="rosrun sub8_missions move_command"
alias araise='rosrun sub8_alarm raise'
alias aclear='rosrun sub8_alarm clear'
alias rsub="export ROS_MASTER_URI='http://192.168.37.60:11311';set_ros_ip"
alias consub="ssh sub8@192.168.37.60 -Y"
alias cm='catkin_make -C ~/mil_ws -j8'
alias killgazebo="killall -9 gazebo & killall -9 gzserver  & killall -9 gzclient"
```

### CODE REVIEW

File: command/sub8_missions/nodes/move_command [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Command Line Mission Runner',
+        usage='Pass any pose editor command with an argument. \n\t forward 1 (moves forward 1 meter) \n\t backward 2ft (moves backward 2 feet), ')
+    parser.add_argument('command', type=str,
+        help='Pose editor command to run')
+    parser.add_argument('argument', type=str, default=0,
+        help='An argument to pass to the command (distance or angle usally). Optionally a unit can be added if a non-standard unit is desired.')
+    parser.add_argument('-m', '--movetype', type=str, default='drive',
+        help='Move type. See lqrrt documentation for info on how to use this.')
```
Comment (DSsoto) [inline command/sub8_missions/nodes/move_command]: This lqrrrt stuff is not supported by the sub currently.

File: command/sub8_missions/tools/waypoint_utility.py [hunk 2]
```diff
@@ -141,6 +135,10 @@ def moveto_action(self, position, orientation):
         rospy.logwarn("Going to waypoint")
         rospy.logwarn("Found server")
 
+        # Stay under the water
+        if position[2] > 0:
+            position[2] = -0.5
```
Comment (DSsoto) [inline command/sub8_missions/tools/waypoint_utility.py:141]: A warning here would be nice

File: command/sub8_missions/nodes/move_command [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    if args.command == 'custom':
+        # Let the user input custom commands, the eval may be dangerous so do away with that at some point.
+        fprint("Moving with the command: {}".format(args.argument))
+        res = yield eval("sub.move.{}.go()".format(args.argument, **action_kwargs))
+
+    elif args.command in ['zrp', 'level_off']:
+        fprint("Zeroing roll and pitch")
+        res = yield sub.move.zero_roll_and_pitch().go(**action_kwargs)
+
+    elif args.command == "stop":
```
Comment (DSsoto) [inline command/sub8_missions/nodes/move_command:45]: Can you add a page to the wiki for all these new subc commands?

**Review by DSsoto — APPROVED** [review]
Looks good, minor comments.

File: command/sub8_missions/missions/tx_test.py [hunk 4]
```diff
@@ -3,4 +3,6 @@
 
 @util.cancellableInlineCallbacks
 def run(sub):
-    yield sub.move.left(3).go()
\ No newline at end of file
+    yield sub.nh.sleep(1)
+    print "Goiung"
```
Comment (kev-the-dev) [inline command/sub8_missions/missions/tx_test.py]: yeah?

File: command/sub8_missions/nodes/move_command [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import numpy as np
+import sub8_tools
+from sub8 import sub_singleton
+from geometry_msgs.msg import PoseStamped, PointStamped
+import missions
+
+import argparse
+
+UNITS = {'m': 1, 'ft': 0.3048, 'yard': 0.9144, 'rad': 1, 'deg': 0.0174533}
+SHORTHAND = {"f": "forward", "b": "backward", "l": "left", "r": "right", "yl": "yaw_left", "yr": "yaw_right"}
```
Comment (kev-the-dev) [inline command/sub8_missions/nodes/move_command]: be nice to add `"d": "down", "u": "up"`

File: command/sub8_missions/nodes/move_command [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    parser.add_argument('command', type=str,
+        help='Pose editor command to run')
+    parser.add_argument('argument', type=str, default=0,
+        help='An argument to pass to the command (distance or angle usally). Optionally a unit can be added if a non-standard unit is desired.')
+    parser.add_argument('-m', '--movetype', type=str, default='drive',
+        help='Move type. See lqrrt documentation for info on how to use this.')
+    parser.add_argument('-t', '--test', action='store_true',
+        help="Runs the mission in test mode and not move the sub.")
+    parser.add_argument('-s', '--speed', type=float, default=0.2,
+        help="How fast to execute the move.")
```
Comment (kev-the-dev) [inline command/sub8_missions/nodes/move_command]: Does this work on the sub? Also, please specify in the help string what this number represents (meters/second or proportion of max speed?)

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
Some quick fixes to add at some point in addition to David's stuff

File: command/sub8_missions/nodes/move_command (hunk 3 above)
Comment (mattlangford) [reply command/sub8_missions/nodes/move_command:45]: Will do

File: command/sub8_missions/nodes/move_command (hunk 6 above)
Comment (mattlangford) [reply command/sub8_missions/nodes/move_command]: Yes, it's m/s (I added that to the help printout)

**Review by kev-the-dev — APPROVED** [review]

**Pull request merged**